tests/test_resolve_command.rs

```diff
@@ -245,9 +245,14 @@ fn check_resolve_produces_input_file(
         test_env.jj_cmd_failure(repo_path, &["resolve", "--config-toml", &merge_arg_config]);
     // This error means that fake-editor exited successfully but did not modify the
     // output file.
-    insta::assert_snapshot!(error, @r###"
-        Error: Failed to use external tool to resolve: The output file is either unchanged or empty after the editor quit.
-        "###);
+    // We cannot use `insta::assert_snapshot!` here after insta 1.22 due to
```

Comment:
Perhaps this function should actually be a macro, then, to preserve the original source locations?

Reply:
I didn't know that was an option. How would I do it?

Reply:
@ilyagr here's an example of calling a macro from another macro: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=071279ba4fa7586c097866dcb314502b . The line/column number seem to be correct.

That being said, passing all four parameters to the macro which `check_resolve_produces_input_file` currently accepts might be unergonomic to use, since the IDE presumably won't help with filling the parameters, so you could leave it as a function instead, at your discretion.

Reply:
the function could return String or bool and assert_snapshot!() or assert!() by caller, if writing macro is too complicated.

Reply:
My guess is that `insta` will only get more confused because of this. It probably expects to be able to locate the snapshot text at this line and column so that `cargo insta review` can edit it. This could be a nice test for it.

Reply:
But overall, it's a neat trick, but complex enough that it doesn't seem worth it just to keep one `assert_snapshot!`.

Reply:
Sorry, I missed your message, @yuja, and now it's too late. 

I did intentionally hide the message inside the function since it's not actually that relevant for the purpose of the function.

Reply:
> I did intentionally hide the message inside the function since it's not actually that relevant for the purpose of the function.

Yep, I agree. So if we want to preserve the source location, `assert!(check_function_that_returns_bool())` is probably better than `assert_snapshot!(string_function(), ...)`.

Reply:
I think that solves a different problem than the one was solving. For what I was doing, once there's no `insta` assertion, it won't panic after upgrading `insta`, and thus there's no problem.

What's the advantage of preserving the source location the way you suggest with a boolean? It seemed to me the only way to figure out which line a test failed at was `RUST_BACKTRACE=1` anyway, but perhaps I missed some other way.

Also, you'd probably want to return a `Result` rather than a `bool` so that you don't lose the information about *why* the `check_function` failed.

Reply:
> I think that solves a different problem

Ah yes. To fix insta compatibility issue, using `assert_eq!()` should be fine.

> What's the advantage of preserving the source location the way you suggest with a boolean?

Suppose `check_resolve_produces_input_file()` is reliable, and one of the three calls failed, an error message like `assertion failed: check_resolve_produces_input_file(..., "right", ...) <file>:<lineno>` might be useful.
OTOH, if we're testing the exact behavior and error message, assert!() inside `check_resolve_produces_input_file()` is probably better.

> Also, you'd probably want to return a Result rather than a bool

Maybe yes if there are multiple reasons to fail, and if that's important.

Reply:
:+1: